client.go

```diff
@@ -533,88 +511,81 @@ func (c *TunaSessionClient) DialWithConfig(remoteAddr string, config *nkn.DialCo
 		return nil, err
 	}
 
-	var lock sync.Mutex
 	var wg sync.WaitGroup
-	conns := make(map[string]*Conn, len(pubAddrs.Addrs))
-	dialer := &net.Dialer{}
 	for i := range pubAddrs.Addrs {
 		wg.Add(1)
 		go func(i int) {
 			defer wg.Done()
 
-			if pubAddrs.Addrs[i] == nil || len(pubAddrs.Addrs[i].IP) == 0 || pubAddrs.Addrs[i].Port == 0 {
-				return
-			}
-
-			netConn, err := dialer.DialContext(ctx, "tcp", fmt.Sprintf("%s:%d", pubAddrs.Addrs[i].IP, pubAddrs.Addrs[i].Port))
-			if err != nil {
-				log.Printf("Dial error: %v", err)
-				return
-			}
-
-			conn := newConn(netConn)
-
-			err = writeMessage(conn, []byte(c.addr.String()), time.Duration(config.DialTimeout)*time.Millisecond)
-			if err != nil {
-				log.Printf("Write message error: %v", err)
-				conn.Close()
-				return
-			}
-
-			metadata := &pb.SessionMetadata{
-				Id: sessionID,
-			}
-			metadataRaw, err := proto.Marshal(metadata)
+			_, err := c.dialAConn(ctx, remoteAddr, sessionID, i, pubAddrs.Addrs[i], config)
 			if err != nil {
-				log.Printf("Encode session metadata error: %v", err)
-				conn.Close()
+				log.Printf("Tuna session dial to ip %v port %v err: %v", pubAddrs.Addrs[i].IP, pubAddrs.Addrs[i].Port, err)
 				return
 			}
-
-			buf, err := c.encode(metadataRaw, remoteAddr)
-			if err != nil {
-				log.Printf("Encode message error: %v", err)
-				conn.Close()
-				return
-			}
-
-			err = writeMessage(conn, buf, time.Duration(config.DialTimeout)*time.Millisecond)
-			if err != nil {
-				log.Printf("Write message error: %v", err)
-				conn.Close()
-				return
-			}
-
-			lock.Lock()
-			conns[connID(i)] = conn
-			lock.Unlock()
 		}(i)
 	}
 	wg.Wait()
 
+	sessKey := sessionKey(remoteAddr, sessionID)
+	c.RLock()
+	conns := c.sessionConns[sessKey]
```

Comment:
What if all dial fails and conns is nil?

Reply:
Yes, modified.